src/lib/notification.vala

```diff
@@ -0,0 +1,254 @@
+/*
+ * This file is part of budgie-desktop
+ *
+ * Copyright © 2018-2022 Budgie Desktop Developers
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ */
+
+ namespace Budgie {
+
+	 /**
+	 * Enumeration of why a notification was closed.
+	 */
+	public enum NotificationCloseReason {
+		/** The notification expired. */
+		EXPIRED = 1,
+		/** The notification was dismissed by the user. */
+		DISMISSED = 2,
+		/** The notification was closed by a call to CloseNotification. */
+		CLOSED = 3,
+		/** Undefined/reserved reasons. */
+		UNDEFINED = 4
+	}
+
+    /**
+     * Enumeration of where notification popups will be shown.
+     */
+    public enum NotificationPosition {
+		TOP_LEFT = 1,
+		TOP_RIGHT = 2,
+		BOTTOM_LEFT = 3,
+		BOTTOM_RIGHT = 4
+	}
+
+	/**
+	 * Enumeration of notification priorities.
+	 */
+	public enum NotificationUrgency {
+		LOW = 0,
+		NORMAL = 1,
+		CRITICAL = 2
+	}
+
+     /**
+	 * This is our wrapper class for a FreeDesktop notification.
+	 */
+	public class Notification : Object {
+		public DesktopAppInfo? app_info { get; private set; default = null; }
+		public string app_id { get; private set; }
+		public NotificationUrgency urgency { get; private set; default = NotificationUrgency.NORMAL; }
+
+		public uint32 id { get; construct; }
+
+		/* Notification information */
+		public string app_name { get; construct; }
+		public HashTable<string, Variant> hints { get; construct; }
+		public string[] actions { get; construct; }
+		public string app_icon { get; construct; }
+		public string body { get; construct set; }
+		public string summary { get; construct set; }
+		public uint expire_timeout { get; construct; }
+
+		public string category { get; construct set; }
+		public int64 timestamp { get; construct set; }
+
+		/* Icon stuff */
+		public Gtk.Image? image { get; set; default = null; }
+
+		private static Regex entity_regex;
+    	private static Regex tag_regex;
+
+		public Notification(
+			uint32 id,
+			string app_name,
+			string app_icon,
+			string summary,
+			string body,
+			string[] actions,
+			HashTable<string, Variant> hints,
+			uint expire_timeout
+		) {
+			Object (
+				id: id,
+				app_name: app_name,
+				app_icon: app_icon,
+				summary: summary,
+				body: body,
+				actions: actions,
+				hints: hints,
+				expire_timeout: expire_timeout
+			);
+		}
+
+		static construct {
+			try {
+				entity_regex = new Regex("&(?!amp;|quot;|apos;|lt;|gt;|#39;|nbsp;)");
+				tag_regex = new Regex("<(?!\\/?[biu]>)");
+			} catch (Error e) {
+				warning("Invalid notificiation regex: %s", e.message);
```

Comment:
notification (not notificiation)

Reply:
Oops. For some reason I have issues spelling 'notification' xD